rts/motoko-rts/src/memory/ic.rs

```diff
@@ -75,26 +77,28 @@ impl Memory for IcMemory {
         let new_hp = old_hp + delta;
 
         // Grow memory if needed
-        grow_memory(new_hp);
+        if (old_hp ^ new_hp) >> 16 != 0 {
+            grow_memory(new_hp)
+        }
 
-        debug_assert!(new_hp <= u64::from(core::u32::MAX));
         HP = new_hp as u32;
 
         Value::from_ptr(old_hp as usize)
     }
 }
 
-/// Page allocation. Ensures that the memory up to, but excluding, the given pointer is allocated.
+/// Page allocation. Ensures that the memory up to, and including, the given pointer is allocated,
+/// with the slight exception of not allocating the extra page for address 0xFFFF_0000.
+/// Enforced precondition: `ptr` must be a pointer less than or equal to 0xFFFF_0000.
 #[inline(never)]
 unsafe fn grow_memory(ptr: u64) {
-    debug_assert!(ptr <= 2 * u64::from(core::u32::MAX));
-    let page_size = u64::from(WASM_PAGE_SIZE.as_u32());
-    let total_pages_needed = ((ptr + page_size - 1) / page_size) as usize;
+    if ptr > 0xFFFF_0000 {
+        // spare the last wasm memory page
+        rts_trap_with("Cannot grow memory")
+    };
+    let total_pages_needed = (ptr >> 16) as usize + (ptr < 0xFFFF_0000) as usize;
```

Comment:
```suggestion
    let total_pages_needed = (ptr >> 16) as usize + (ptr < 0xFFFF_0000) as usize;
```
Why the conditional +1? 


Can we add a debug_assert that the memory_size is < 65536?

Reply:
The `+1` is explained in the comment above.

Yes, sure I can add an assert.